### ISSUE TITLE
False positive: prefer_self_in_static_references with return types using collection type literals in a class

### ISSUE THREAD

**Issue by ainame** [realm/SwiftLint#5042]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Since SwiftLint 0.52.0, I started seeing SwiftLint pointing `prefer_self_in_static_references` in the wrong places. For example, this `TestClass` triggers `prefer_self_in_static_references` for return types.

```swift
class TestClass {
    required init() {}

    // Trigger `Prefer Self in Static References Violation: Use `Self` to refer to the surrounding type name (prefer_self_in_static_references)`
    static func makeArray() -> [TestClass] {
        []
    }

    //  Trigger `Prefer Self in Static References Violation: Use `Self` to refer to the surrounding type name (prefer_self_in_static_references)`
    static func makeDictionary() -> [String: TestClass] {
        [:]
    }
}
```

But after following the suggestion to use `Self`, this can't be compiled with Xcode 14.3. 

```swift
class TestClassWithSelf {
    required init() {}

    // Compilation error `Covariant 'Self' or 'Self?' can only appear at the top level of method result type`
    static func makeArray() -> [Self] {
        []
    }

    // Compilation error `Covariant 'Self' or 'Self?' can only appear at the top level of method result type`
    static func makeDictionary() -> [String: Self] {
        [:]
    }
```

### Environment

* SwiftLint version: 0.52.2 (this started since 0.52.0)
* Installation method used: CocoaPods
* Paste your configuration file:

```yml
opt_in_rules:
  - prefer_self_in_static_references
```

* Xcode: 14.3
